### PR TITLE
Rewrite Leap and AttackLeap, add LeapAttack and EdibleByAttackLeap

### PR DESCRIPTION
**Pull request by abcdefg30**
Yet another attempt at resolving the `AttackLeap` mess. Supersedes #13513.

### CODE REVIEW

**Comment by GraionDilach** [comment]
Something I didn't understand back at #13513 either... if you remove the attacking angle, why are you still calling this a leap? This becames a jousting attack then.

**Comment by pchote** [comment]
`AttackPoke`? :smile:

**Comment by GraionDilach** [comment]
Well, there will be need for a variant with the angle being kept for RA2 atleast longterm whicchs hould have the `AttackLeap` name, so I'm fine with `AttackPoke`/`AttackJoust`/etc.

**Comment by abcdefg30** [comment]
I'll go with `AttackJoust` then if noone objects.

**Comment by Smittytron** [comment]
Played around with this. Naming conventions aside, this is fixes the dog issues I know of. I doubt we'll see unwelcome balance implications here but that's something we can address later as needed.

**Comment by UsernameNotSure** [comment]
Hi, just registered to outline the issues on this again.
The dog leap logic has many individual factors that you most likely DON'T want to put into a generalized, single design.

The issues:
1) Pathfinding yes / no:
If the leap range is a huge amount, let's say 9 fields, dogs would be able to jump through structures, over sandbags / walls, over rivers, up / down cliffs -> line of fire check for the weapon in terms of passable terrain to (dis-)allow such jumps
2) Instakill yes / no:
Normally dogs should instakill other dogs and infantry. Eventually that'll change one day. So it should be possible to define whether to use the dog's "weapon damage" or a plain instakill. Eventually also with individual damage tables vs certain units, such as Tanya to get jumped but not killed instantly (and such actions would then instead result in the dog dealing damage to the target unit but being killed by it).
3) two dogs jumping each other simultaneously:
Before both ended up in an endless loop of jumping another. Workarounds could be to randomly assign one dog that gets jumped and the other one doing the jump. As both should die anyway a new animation would be required in which the dogs kill another - unless the in "2)" desribed damage table was used in which case the jumped would receive damage but end up killing the jumping unit.
4) many dogs, same target:
If multiple dogs are assigned the same target their unit-target order should become a ground-target order once the 1st dog leaps at the target and in general any target that's dog-leaped shouldn't be targeted by other dogs anymore. Otherwise it'd be possible that many dogs pile up on the same x,y coords, standing inside another or eventually kill each other if their projectile may cause that.

**Comment by abcdefg30** [comment]
1. is out of scope for this PR.
2. - 4. are solved in this PR, dogs deal the damage defined on the warhead, don't loop endlessly and also don't pile up on the same targets location.

**Comment by Smittytron** [comment]
AttackLunge?

**Comment by GraionDilach** [comment]
Again, I am fine with anything as long as it's not kept as Leap, since there is no leap anymore but this old AttackLeap will have to be hacked into the RA2 repo, following the bag handling's footsteps and it feels weird to come up with a name for the old logic because the new logic has the old name as an artifact.

**Comment by pchote** [comment]
It would be good to get this in for the playtest, so can you proceed with the rename @abcdefg30.  This is breaking compatibility with the old trait behaviour, so we will need some form of update rule too.  I suggest (assuming the rename goes ahead) automatically removing the `AttackLeap` trait and giving a manual step saying that that trait has been removed, and to define new attack logic using the `AttackBite` (or whatever) trait.

File: mods/ra/rules/infantry.yaml
Comment (pchote) [inline mods/ra/rules/infantry.yaml]: Wouldn't it make more sense to use one of the non-Infantry sprite bodies for these?

**Comment by pchote** [comment]
It looks like it might make sense to rebase this on top of #15140? We may be able to use `GrantConditionOnArmamentReloading` to grant the `rip` condition and the `INotifyAiming` interface (via a new `GrantConditionWhileAiming` trait) to grant the `run` condition?

**Comment by abcdefg30** [comment]
Makes sense, let's try that.

**Comment by pchote** [comment]
#15140 has now been merged.

**Comment by pchote** [comment]
It sounds like @abcdefg30 won't have time to work on this again soon, so dropping from the milestone so we can focus on the other areas.

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionWhileAiming.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionWhileAiming.cs]: Thinking out loud, maybe a job for a different PR: if `INotifyAiming` provided some information about *what* was being targeted then we could add a filter (probably using target types), which would solve the plumbing for #2125.

File: OpenRA.Mods.Cnc/Traits/Attack/AttackLeap.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/Attack/AttackLeap.cs]: `INotifyCreated` can be dropped here (it comes in via inheritance)

**Comment by pchote** [comment]
We get some fairly ugly position mismatches when the dog attacks an infantry while it is moving between cells.  I believe we can fix this by having:
1. `Leap` set the underlying cell/subcell as it does now
2. track the `target.CenterPosition` as it moves each tick (updating VisualPosition)
3. play the rip animation at whatever position (in-cell or not) the target was at on impact
4. Visually move the dog to the center of the target subcell once the rip animation is complete

This would then match the behaviour we have with docking, C4, and movement.

**Comment by abcdefg30** [comment]
Updated.

**Comment by GraionDilach** [comment]
No renaming?

**Comment by abcdefg30** [comment]
Yes. That can be left for a follow-up PR.

**Review by pchote — COMMENTED** [review]
Just a couple of comments/questions, otherwise LGTM. Seems to behave well ingame.

**Review by Smittytron — DISMISSED** [review]
Tested this with dog-heavy missions like Soviet 03 and Soviet 07. I don't think we'll need to re-balance anything.
New dogs look great.

**Review by pchote — DISMISSED** [review]

**Comment by pchote** [comment]
@reaperrr is this going to conflict with #15675 too much?

**Comment by reaperrr** [comment]
> @reaperrr is this going to conflict with #15675 too much?

There's going to be a lot of merge conflicts, but one of our PRs was bound to have to go through that. Not worth delaying this PR over.

File: OpenRA.Mods.Cnc/Traits/EdibleByLeap.cs
Comment (reaperrr) [inline OpenRA.Mods.Cnc/Traits/EdibleByLeap.cs]: Maybe I'm missing something, but why do we return 'true' if the leaper is null or dead?

Shouldn't this be `leaper != null && !leaper.IsDead && leaper == targeter`?

**Comment by reaperrr** [comment]
👍 
Only two tiny issues found: `LeapAttack` was in Common namespace instead of Cnc, and the last commits' title was a bit too long.
I didn't want to delay the merge over such minor issues, so I pushed the fixes myself and will merge this after Travis approves.

**Review by reaperrr — DISMISSED** [review]

**Review by reaperrr — APPROVED** [review]

**Pull request merged**
